src/audio/kpb.c

```diff
@@ -0,0 +1,541 @@
+		return NULL;
+	}
+
+	cd->history_depth = ipc_kpb->history_depth;
+	comp_set_drvdata(dev, cd);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+
+}
+
+/**
+ * kpb_free() - reclaim memory for a key phrase buffer.
+ * @arg1: component device pointer
+ */
+static void kpb_free(struct comp_dev *dev)
+{
+	struct comp_data *kd = comp_get_drvdata(dev);
+
+	trace_kpb("kpb_free()");
+
+	rfree(kd);
+	rfree(dev);
+}
+
+static int kpb_trigger(struct comp_dev *dev, int cmd)
+{
+	trace_kpb("kpb_trigger()");
+
+	return comp_set_state(dev, cmd);
+}
+
+/**
+ * kpb_prepare() - prepare key phrase buffer.
+ * @arg1:  kpb component.
+ *
+ * Return: integer representing either 0 - success
+ * or -EINVAL - failure.
+ */
+static int kpb_prepare(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int ret = 0;
+	int i;
+
+	trace_kpb("kpb_prepare()");
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+	if (ret < 0)
+		return ret;
+
+	cd->no_of_clients = 0;
+
+	/* allocate history_buffer/s */
+#if KPB_NO_OF_HISTORY_BUFFERS == 2
+
+	cd->his_buf_hp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+					  MAX_BUFFER_SIZE - LPSRAM_SIZE);
+	cd->his_buf_lp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_LP,
+					  LPSRAM_SIZE);
+
+	if (cd->his_buf_hp.sta_addr || cd->his_buf_lp.sta_addr) {
+		trace_kpb_error("Failed to allocate space for "
+				"KPB bufefrs");
+		return -EINVAL;
+	}
+
+	cd->his_buf_hp.end_addr = cd->his_buf_hp.sta_addr +
+	(MAX_BUFFER_SIZE - LPSRAM_SIZE);
+
+	cd->his_buf_lp.end_addr = cd->his_buf_lp.sta_addr +
+	LPSRAM_SIZE;
+
+#elif KPB_NO_OF_HISTORY_BUFFERS == 1
+
+	cd->his_buf_hp.sta_addr = rballoc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+					  MAX_BUFFER_SIZE);
+
+	if (cd->his_buf_hp.sta_addr) {
+		trace_kpb_error("Failed to allocate space for "
+				"KPB bufefrs");
+	return -EINVAL;
+	}
+
+	cd->his_buf_hp.end_addr = cd->his_buf_hp.sta_addr + MAX_BUFFER_SIZE;
+
+#else
+#error "Wrong number of key phrase buffers configured"
+#endif
+
+	/* TODO: zeroes both buffers */
+
+	/* Initialize clients data */
+	for (i = 0; i < MAX_NO_OF_CLIENTS; i++)
+		cd->clients[i].state = KPB_CLIENT_UNREGISTERED;
+
+	return ret;
+}
+
+/**
+ * kpb_copy() - copy real time input stream into sink buffer,
+ * and in the same time buffers that input for
+ * later use by some of clients.
+ *
+ * @arg1:  kpb component.
+ *
+ * Return: integer representing either 0 - success
+ * or -EINVAL - failure.
+ */
+static int kpb_copy(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *kpb = comp_get_drvdata(dev);
+	struct comp_buffer *source;
+	struct comp_buffer *sink;
+
+	trace_kpb("kpb_copy()");
+
+	/* Get source and sink buffers */
+	source = list_first_item(&dev->bsource_list, struct comp_buffer,
+				 source_list);
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
+			       sink_list);
+	/* Process source data */
+	/* Check if source have data and valid pointer to read from */
+	if (source->avail > 0 && source && sink && source->r_ptr) {
```

Comment:
Checking source after getting source->avail doesn't make any sense.
Also r_ptr and w_ptr are initialized on buffer creation, so I think checking them is not necessary.

Reply:
For the first sentence I agree, for the second, I would leave this check of r_ptr & w_ptr as it is - at least for cmocka tests.